test_conformance/subgroups/test_workgroup.cpp

```diff
@@ -2022,31 +2008,106 @@ struct SCIN_NU {
+            for (j = 0; j < nj; ++j) {      // for each subgroup
+                ii = j * ns;
+                n = ii + ns > nw ? nw - ii : ns;
+
+                for (i = 0; i < n; ++i)
+                    t[ii + i] = (Ty)i;
+            }
+
+            // Now map into work group using map from device
+            for (j = 0; j < nw; ++j) {
+                i = m[4 * j + 1] * ns + m[4 * j];
+                x[j] = t[i];
+            }
+
+            x += nw;
+            m += 4 * nw;
+        }
+    }
+
+    static int chk(Ty* x, Ty* y, Ty* mx, Ty* my, cl_int* m, int ns, int nw, int ng)
+    {
+        int ii, i, j, k, n;
+        int nj = (nw + ns - 1) / ns;
+        Ty tr, rr, trt;
+
+        log_info("  sub_group_non_uniform_scan_inclusive_%s(%s)...\n", Which == 4 ? "and" : (Which == 5 ? "or" : "xor"), TypeName<Ty>::val());
+
+        for (k = 0; k < ng; ++k) {      // for each work_group
+            // Map to array indexed to array indexed by local ID and sub group
+            for (j = 0; j < nw; ++j) {  // inside the work_group
+                i = m[4 * j + 1] * ns + m[4 * j];
+                mx[i] = x[j];           // read host inputs for work_group
+                my[i] = y[j];           // read device outputs for work_group
+            }
+
+            for (j = 0; j < nj; ++j) {  // for each subgroup
+                ii = j * ns;
+                n = ii + ns > nw ? nw - ii : ns;
+                // Check result
+                for (i = 0; i < n; ++i) {   // inside the subgroup
+                    if (i > NON_UNIFORM - 1) {
+                        tr = 0;
+                    }
+                    else {
+                        if (Which == 4) {  // function and
+                            tr = i == 0 ? mx[ii] : tr & mx[ii + i];
                         }
                         else if (Which == 5) {  // function or
-                            //tr = i == 0 ? mx[ii] : tr | mx[ii + i];
-                            tr = i == 0 ? mx[ii] : val1 | val2;
+                            tr = i == 0 ? mx[ii] : tr | mx[ii + i];
                         }
                         else if (Which == 6) {  // function xor
-                            //tr = i == 0 ? mx[ii] : tr ^ mx[ii + i];
-                            tr = i == 0 ? mx[ii] : val1 ^ val2;
+                            tr = i == 0 ? mx[ii] : tr ^ mx[ii + i];
                         }
                         else {
-                            log_error("ERROR: sub_group_non_unifor_scan_inclusive - unknown function type number");
+                            log_error("ERROR: sub_group_non_uniform_scan_inclusive - unknown function type number");
                             return -1;
                         }
 
                     }
                     trt = mx[ii + i];
                     rr = my[ii + i];
-                    std::memcpy(&val1, &tr, size_to_copy);
-                    std::memcpy(&val2, &trt, size_to_copy);
```

Comment:
Wouldn't moving these `memcpy`s earlier do the trick? I'm concerned about code duplication.

Reply:
Moving memcpy up wouldn't do the trick, as this entire logic is incorrect. `size_to_copy` variable is set only if `Ty` is `double`, `float `or `half`.  `size_to_copy` variable is used to copy the values into `val1` and `val2` variables - those variables are used only for logical operations (and, or, xor) which do not operate on floating point types, so `Ty` is never floating point - it implicates that `size_to_copy` is never set.

For me, it doesn't make sense to put arithmetic operations (add, max, min, mul) and logical operations (and, or, xor) in the same template, as those two groups operate on different types of data. The second one does not operate on floating point. The code logic above (with memcpy) is just a hack for me and it's not readable. That's why I think that a little code duplication is reasonable in this case as it makes the code easier to read and not-hacky.

Reply:
I mostly agree, but the `memcpy` hack is already used in other templates. I'm okay to make this fix as-is, but for the maintability and readability, the operations could be implemented as function templates, so that bitwise operations wouldn't be instantiated for floats and the `memcpy` hack could be avoided. @gwawiork